jupyterhub/proxy.py

```diff
@@ -300,14 +300,27 @@ async def add_user(self, user, server_name='', client=None):
             spawner.server.host,
             {'user': user.name, 'server_name': server_name},
         )
+        await self.add_route(
+            url_path_join('/socket/', spawner.proxy_spec),
+            '{}://{}:3000'.format(spawner.server.proto, spawner.server.ip),
+            {'user': user.name, 'server_name': server_name, 'target': 'socket'},
+        )
+        await self.add_route(
+            url_path_join(spawner.proxy_spec, '/logout'),
+            url_path_join(spawner.public_domain, '/hub/logout'),
+            {'user': user.name, 'server_name': server_name, 'target': 'logout'},
+        )
 
     async def delete_user(self, user, server_name=''):
         """Remove a user's server from the proxy table."""
         routespec = user.proxy_spec
         if server_name:
             routespec = url_path_join(user.proxy_spec, server_name, '/')
         self.log.info("Removing user %s from proxy (%s)", user.name, routespec)
+        self.log.info("user proxyspec: %s", user.proxy_spec)
         await self.delete_route(routespec)
+        await self.delete_route(url_path_join('/socket/', user.proxy_spec))
+        await self.delete_route(url_path_join(user.proxy_spec, '/logout'))
```

Comment:
I think you forgot here `/hub/logout`. You could write small getter functions that construct there URLs such that you do not need to duplicate them

Reply:
`hub/logout` was not forgotten. It is the url where a user logs out of the hub and is not specific to a user. As such we should never remove it while the hub is running. Here we delete the route that points from the code-server logout point to the hub logout and to delete that route we do not need to specify the target of the route.